oeps/oep-0041-arch-async-server-event-messaging.rst

```diff
@@ -0,0 +1,486 @@
+
+Reverse DNS
+~~~~~~~~~~~
+Example: ``org.openedx``
+
+In this example, ``org.openedx`` is the generic prefix that would be used for
+all Open edX installs. Events that are strictly edX-specific, like those that
+might interact with internal IT or finance reporting systems, should use
+``org.edx`` as the prefix instead. If in doubt, default to ``org.openedx``.
+
+Domain (from Domain Driven Design)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Example: ``catalog``
+
+The next part of the ``type`` hierarchy is the Domain. Examples of this are:
+
+* catalog
+* content_authoring
+* credentials
+* ecommerce
+* learning
+
+It is the expectation that there are relatively few domains, and that they will
+roughly match deployed services. Domain names should be lower cased and use
+underscores if they are more than one word.
+
+Subject
+~~~~~~~
+Example: ``course``
+
+The name of an entity that the event applies to. Examples might be ``course``,
+``student``, ``enrollment``, ``order``, etc. Subjects may be namespaced, so
+``special_exam.proctored.allowance`` could be a subject.
+
+A subject should always mean the same thing within a Domain, but can mean
+different things across domains. For instance, what the LMS (``learning``
+domain) calls a ``course`` might map to what the ``catalog`` domain would call a
+``course_run``. We should try to be consistent where possible, but each domain
+ultimately gets to decide what its terms mean, and we should be careful when
+translating a concept from one domain to another. For instance, the
+``content_authoring`` and ``learning`` domains might both have a concept of a
+"due date" for an assignment. But while the ``content_authoring`` due date is
+determined only by the content author, the ``learning`` due date might take into
+account a student's cohort, individual due date extensions, accessibility
+allowances, and any number of other things. Both domains may call it ``due``,
+but the due date information from ``content_authoring`` is just an input to the
+more complex due date information in ``learning``.
+
+Action
+~~~~~~
+Example: ``created``
+
+This is the action that occurred for the event. Some of most common ones will be
+``created``, ``updated``, and ``deleted``, but many applications will want more
+specific actions like ``declined``, ``started``, ``verified``, etc. Applications
+should prefer these more specific actions when possible, since they are usually
+clearer for listeners and don't requiring knowing implementation details. For
+instance, it is preferable to have a ``registration.completed`` event than to
+have a ``registration.updated`` event and leave it to clients to check the
+resulting message's payload for ``status: "complete"``.
+
+Actions should be past tense, to better align with our existing conventions
+around Django signals and learning analytics events (we're not completely
+consistent, but we tend towards past tense).
+
+Major Version
+~~~~~~~~~~~~~
```

Comment:
Question about versioning... I think we need to make sure we outline a process for deprecating old major message versions. I'm imagining a future where we're on V4 of some message in the LMS, but still need to send messages V1 2 and 3 (otherwise we break someone's consumer), resulting in us sending a flurry of messages all the time something something queue size something something memory limit :)

Reply:
I agree, I need to really flesh out the versioning story here, especially given how granular these contracts effectively are since individual messages have versions.

That being said (and I hope these words don't bite me later), I don't think we should be too concerned about writing v1, 2, 3, 4 of a message. Backwards compatible minor version bumps should be preferred, of course, but writing multiple major versions of a messages grows slowly and linearly over time. I'm hoping that simple event emission is less of a dev support burden than maintaining bidirectional APIs. My big concern for message spew are if we do things where we the number of messages sent grows with the size of the thing being worked on, for instance if we have some action on the course that would unenroll a thousand students, and we were trying to send all those messages in a synchronous transaction.

I honestly hadn't really thought about that before now. That could really bite us if we start tying messages to model level interactions. Argh. Thank you.